src/include/types.h

```diff
@@ -380,48 +380,76 @@ inline ostream& operator<<(ostream& out, const prettybyte_t& b)
+      const int index, const uint64_t mult, const char* u)
+  {
+    char buffer[32];
+
+    if (index == 0) {
+      (void) snprintf(buffer, sizeof(buffer), "%" PRId64, n);
+    } else if ((v & (mult - 1)) == 0) {
+      // If this is an even multiple of the base, always display
+      // without any decimal fraction.
+      (void) snprintf(buffer, sizeof(buffer), "%" PRId64 "%s", n, u);
+    } else {
+      // We want to choose a precision that reflects the best choice
+      // for fitting in 5 characters.  This can get rather tricky when
+      // we have numbers that are very close to an order of magnitude.
+      // For example, when displaying 10239 (which is really 9.999K),
+      // we want only a single place of precision for 10.0K.  We could
+      // develop some complex heuristics for this, but it's much
+      // easier just to try each combination in turn.
+      int i;
+      for (i = 2; i >= 0; i--) {
+        if (snprintf(buffer, sizeof(buffer), "%.*f%s", i,
+          static_cast<double>(v) / mult, u) <= 5)
+          break;
+      }
+    }
+
+    return out << buffer;
+  }
+}
+
+struct dec_u_t {
   uint64_t v;
-  // cppcheck-suppress noExplicitConstructor
-  si_t(uint64_t _v) : v(_v) {}
+  explicit dec_u_t(uint64_t _v) : v(_v) {};
 };
 
-inline ostream& operator<<(ostream& out, const si_t& b)
+inline ostream& operator<<(ostream& out, const dec_u_t& b)
 {
-  char buffer[32];
   uint64_t n = b.v;
   int index = 0;
+  const char* u[] = {" ", "k", "M", "G", "T", "P", "E"};
 
-  while (n >= 1024 && index < 6) {
+  while (n >= 1024 && index < 7) {
     n /= 1024;
     index++;
   }
 
-  char u = " KMGTPE"[index];
+  return format_u(out, b.v, n, index, 1ULL << (10 * index), u[index]);
+}
 
-  if (index == 0) {
-    (void) snprintf(buffer, sizeof(buffer), "%" PRId64, n);
-  } else if ((b.v & ((1ULL << 10 * index) - 1)) == 0) {
-    // If this is an even multiple of the base, always display
-    // without any decimal fraction.
-    (void) snprintf(buffer, sizeof(buffer), "%" PRId64 "%c", n, u);
-  } else {
-    // We want to choose a precision that reflects the best choice
-    // for fitting in 5 characters.  This can get rather tricky when
-    // we have numbers that are very close to an order of magnitude.
-    // For example, when displaying 10239 (which is really 9.999K),
-    // we want only a single place of precision for 10.0K.  We could
-    // develop some complex heuristics for this, but it's much
-    // easier just to try each combination in turn.
-    int i;
-    for (i = 2; i >= 0; i--) {
-      if (snprintf(buffer, sizeof(buffer), "%.*f%c", i,
-        (double)b.v / (1ULL << 10 * index), u) <= 5)
-        break;
-    }
+struct bin_u_t {
+  uint64_t v;
+  explicit bin_u_t(uint64_t _v) : v(_v) {};
+};
+
+inline ostream& operator<<(ostream& out, const bin_u_t& b)
+{
+  uint64_t n = b.v;
+  uint64_t mult = 1;
+  int index = 0;
+  const char* u[] = {" ", "KiB", "MiB", "GiB", "TiB", "PiB", "EiB"};
```

Comment:
hmm, is this always going to be B (bytes)?  I guess so?

Reply:
yes ```bin_u_t``` should only be used for bits and bytes. I haven't encountered a place where bits are printed so adding the unit seemed reasonable. Will add comments to the structs.